src/mon/OSDMonitor.cc

```diff
@@ -1625,18 +1625,21 @@ bool OSDMonitor::can_mark_in(int i)
   return true;
 }
 
-void OSDMonitor::check_failures(utime_t now)
+bool OSDMonitor::check_failures(utime_t now)
 {
+  bool propose_now = false;
   for (map<int,failure_info_t>::iterator p = failure_info.begin();
        p != failure_info.end();
        ++p) {
     if (can_mark_down(p->first)) {
-      check_failure(now, p->first, p->second);
+      check_failure(now, p->first, p->second, propose_now ? NULL : &propose_now);
     }
   }
+  return propose_now;
```

Comment:
i think usually we can finished the proposal within one tick(5s by default IIRC), if not, it would be OK(or better) to trigger the proposal again. Make sense?